NTC multi TS : loading from new format

guilpier-code opened the pull request.

guilpier-code committed:
NTC multi TS - save into new format : version --> 820 + refactoring links save (split save in 2 parts)

guilpier-code committed:
NTC multi TS - save into new format : actual save

guilpier-code committed:
NTC multi TS : load from new format

guilpier-code committed:
[skip ci] NTC multi TS - load from new format : some corrections due to sonarQube review

flomnes committed:
Merge pull request #458 from AntaresSimulatorTeam/develop

Merge develop

flomnes committed:
Merge remote-tracking branch 'http/feature/multiple-ntc-timeseries' into feature/ntc-multi-ts--load

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -395,21 +395,56 @@ bool linkLoadTimeSeries_version_320_to_630(AreaLink& link, const AnyString& fold
-            ret = link.loadDataFromCSVfile(Matrix<>::optFixedSize | Matrix<>::optImmediate) && ret;
-*/
-bool linkLoadTimeSeries_version_630_to_820(AreaLink& link, const AnyString& folder)
+static bool linkLoadTimeSeries_version_630_to_820(AreaLink& link, const AnyString& folder)
```
`linkLoadTimeSeries_version_630_to_810`, since format will change at 820, we want to exclude it. Also, add the "legacy" keyword to this function and the other 2 legacy functions.

guilpier-code replied:
The first part of the request above was done.
I think the "legacy" keyword is useless, as loading functions have a name expressive enough.

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -395,21 +395,56 @@ bool linkLoadTimeSeries_version_320_to_630(AreaLink& link, const AnyString& fold
+    return true;
+}
+
+static bool linkLoadTimeSeries(AreaLink& link, const AnyString& folder)
```
linkLoadTimeSeries_versions_820_and_later

guilpier-code replied:
ok, done

flomnes commented on src/solver/constraints-builder/cbuilder.cpp:
```diff
@@ -91,9 +91,28 @@ bool CBuilder::update(bool applyCheckBox)
     {
-        // Try to open the file
-        if (not (*linkInfoIt)->ptr->loadDataFromCSVfile(Matrix<>::optImmediate))
+        auto linkInfo = *linkInfoIt;
+        Data::AreaLink* link = linkInfo->ptr;
+
+        // Try to open link data files
+        // ... Load parameters from a data file 
+        YString dataFilename = link->parameters.jit->sourceFilename;
+        if (not link->parameters.loadFromCSVFile(dataFilename, Data::fhlMax, HOURS_PER_YEAR, Matrix<>::optImmediate))
+        {
+            return false;
+        }
+        // ... Load direct capacities from a data file 
+        dataFilename = link->directCapacities.jit->sourceFilename;
+        if (not link->directCapacities.loadFromCSVFile(dataFilename, Data::fhlMax, HOURS_PER_YEAR, Matrix<>::optImmediate))
+        {
+            return false;
+        }
+        // ... Load indirect capacities from a data file 
+        dataFilename = link->indirectCapacities.jit->sourceFilename;
+        if (not link->indirectCapacities.loadFromCSVFile(dataFilename, Data::fhlMax, HOURS_PER_YEAR, Matrix<>::optImmediate))
+        {
```
This will fail on older studies. Require study upgrade before running the Kirchoff generator ?

flomnes replied:
Or maybe check the study version number to use the legacy format when necessary ?

flomnes replied:
Create common loading function used also in `CBuilder::deletePreviousConstraints`

guilpier-code replied:
ok, done

guilpier-code committed:
NTC with multi time series : corrections due to automatic constraints generator incorrectness

- We embed the loading of time-series for any version of the study in a link class public method.
- Then we can use it to load links in the automatic constraints generator more easily

flomnes commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -72,6 +72,25 @@ AreaLink::~AreaLink()
+bool AreaLink::loadTimeSeries(Study& study, const AnyString& folder)
+{
+    if (study.header.version < 320)
+        return linkLoadTimeSeries_for_version_under_320(*this, folder, study);
+    else if (study.header.version < 630)
+        return linkLoadTimeSeries_for_version_from_320_to_630(*this, folder);
+    else if (study.header.version < 820)
+        return linkLoadTimeSeries_for_version_from_630_to_810(*this, folder);
+    else {
+        return linkLoadTimeSeries_for_version_820_and_later(*this, folder);
```
Since you pass `*this`, it seems natural to make all these functions private members of class `AreaLink`.

guilpier-code replied:
ok, done

flomnes reviewed (commented):
LGTM, except for one minor detail, see comment.

flomnes committed:
Merge branch feature/multiple-ntc-timeseries into ntc-multi-ts--load

guilpier-code committed:
NTC load TS : corrections due to code review

making loading functions methods of class AreaLink
